src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.OSX/CertificateData.cs

```diff
@@ -59,126 +61,57 @@ internal CertificateData(byte[] rawData)
 
-            Issuer = new X500DistinguishedName(tbsCertificate.ReadNextEncodedValue());
+            NotBefore = tbsCertificate.Validity.NotBefore.GetValue().UtcDateTime;
+            NotAfter = tbsCertificate.Validity.NotAfter.GetValue().UtcDateTime;
 
-            DerSequenceReader validity = tbsCertificate.ReadSequence();
-            NotBefore = validity.ReadX509Date();
-            NotAfter = validity.ReadX509Date();
+            Subject = new X500DistinguishedName(tbsCertificate.Subject.ToArray());
 
-            if (validity.HasData)
-                throw new CryptographicException(SR.Cryptography_Der_Invalid_Encoding);
-
-            Subject = new X500DistinguishedName(tbsCertificate.ReadNextEncodedValue());
-
-            SubjectPublicKeyInfo = tbsCertificate.ReadNextEncodedValue();
-            DerSequenceReader subjectPublicKeyInfo = new DerSequenceReader(SubjectPublicKeyInfo);
-            DerSequenceReader subjectKeyAlgorithm = subjectPublicKeyInfo.ReadSequence();
-            PublicKeyAlgorithm.AlgorithmId = subjectKeyAlgorithm.ReadOidAsString();
-            PublicKeyAlgorithm.Parameters = subjectKeyAlgorithm.HasData ? subjectKeyAlgorithm.ReadNextEncodedValue() : Array.Empty<byte>();
-
-            if (subjectKeyAlgorithm.HasData)
-                throw new CryptographicException(SR.Cryptography_Der_Invalid_Encoding);
-
-            PublicKey = subjectPublicKeyInfo.ReadBitString();
-
-            if (subjectPublicKeyInfo.HasData)
-                throw new CryptographicException(SR.Cryptography_Der_Invalid_Encoding);
-
-            if (Version > 0 &&
-                tbsCertificate.HasData &&
-                tbsCertificate.PeekTag() == DerSequenceReader.ContextSpecificConstructedTag1)
-            {
-                IssuerUniqueId = tbsCertificate.ReadBitString();
-            }
-            else
+            using (AsnWriter writer = AsnSerializer.Serialize(tbsCertificate.SubjectPublicKeyInfo, AsnEncodingRules.DER))
             {
-                IssuerUniqueId = null;
+                SubjectPublicKeyInfo = writer.Encode();
             }
 
-            if (Version > 0 &&
-                tbsCertificate.HasData &&
-                tbsCertificate.PeekTag() == DerSequenceReader.ContextSpecificConstructedTag2)
-            {
-                SubjectUniqueId = tbsCertificate.ReadBitString();
-            }
-            else
-            {
-                SubjectUniqueId = null;
-            }
+            PublicKeyAlgorithm.AlgorithmId = tbsCertificate.SubjectPublicKeyInfo.Algorithm.Algorithm.Value;
+            PublicKeyAlgorithm.Parameters = tbsCertificate.SubjectPublicKeyInfo.Algorithm.Parameters?.ToArray() ?? Array.Empty<byte>();
+
+            PublicKey = tbsCertificate.SubjectPublicKeyInfo.SubjectPublicKey.ToArray();
+
+            IssuerUniqueId = Version >= 1 ? tbsCertificate.IssuerUniqueId?.ToArray() : null;
+            SubjectUniqueId = Version >= 1 ? tbsCertificate.SubjectUniqueId?.ToArray() : null;
```

Comment:
Most of the fields should be dropped in favor of saving the serializer output.  These two would maintain consistency with current code by saying that if `Version < 1` and they aren't `null` to throw the invalid encoding exception.

Reply:
Agreed, in fact the whole class should be dropped (see comment above) in favor of using `CertificateAsn` directly. It doesn't offer any additional value except few checks and cached Issuer+Subject.